docs/madr/decisions/057-rework-mesh-routes-matching.md

```diff
@@ -0,0 +1,826 @@
+
+### Add possibility to put Mesh*Routes in `spec.to[].targetRef`
+
+#### Advantages
+
+- simplifying policy api as `spec.targetRef` will always pick proxy and `spec.to[].targetRef` section will select outgoing traffic
+- ability to set different confs on different proxy for the same route (overriding default conf for route without producer/consumer policies, on universal)
+- gives more flexibility as you can target multiple outbounds in single policy. Reducing number of consumer policies
+- more consistent with current approach
+- easier policy matching code
+- least problematic migration
+
+#### Disadvantages
+
+- can be harder to understand because `spec.targetRef` can be MeshSubset (which is optional and probably will rarely be used)
+- yet another migration for users (only for system policies as for namespaced policies this is already not allowed) 
+- mixing outbounds and routes in `spec.to[].targetRef`
+- hides the most often most important information under `to[]` adding noise when targeting Mesh in `spec.targetRef`
+- is it consistent in the case of MeshGateways? MeshHTTPRoutes aren't on outbound listeners in the case of MeshGateways, they're on inbound listeners
+
+### Leave Mesh\*Routes in top level targetRef as it is now and refine semantics of `spec.targetRef`
+
+Instead of having `targetRef` always refer to a workload,
+instead `targetRef` selects some part of the Envoy configuration of a workload,
+where that part is represented by a real Kuma resource.
+
+In fact `targetRef` is the wrong concept for selecting workloads,
+since "targetRef refers to a workload" means we aren't targeting
+an arbitrary `Kind` of resource but instead always a `Dataplane`.
+
+In order to allow limiting the effects of a policy further
+we can introduce `spec.workloadSelector` which would basically be some variation of `labelSelector`.
+
+#### Advantages:
+
+- the workloads targeted by the policy is often "all workloads" or can be
+  inferred from the namespace so we kind of waste the top level, "special" `spec.targetRef`
+- tends to be less noisy as we move most interesting part of the policy from `to` /`from`section
+- Doesn't force every potential resource into a "to" or a "from", instead using the semantics of the `kind` itself
+	- `HTTPRoute` is a route, regardless of whether it's inbound or outbound
+	- `MeshService` _is_ a destination so `spec.targetRef.kind: MeshService` targets the configuration for a given destination
+	- `WorkloadIdentity` (obviously doesn't exist yet but we want to change how we manage identity)
+      would represent an identity so fits as `spec.targetRef` for e.g. `MeshTrafficPermission`
+
+#### Disadvantages:
+
+- `targetRef` is a single field, so it's not easy to share configuration
+  in a single resource to avoid duplication.
+  Though the very WIP policy attachment GEPs do propose a `targetRefs` field
+- requires a huge migration
```

Comment:
can you add more details? in top level targetRef we only have `Mesh*Route` right? It's not like we have every policy there, right?

Reply:
> in top level targetRef we only have Mesh*Route right? 

Yes, and the problem is thus all the outbound policies would need to change, where we have `spec.to[].targetRef.kind: MeshService`.